### PR TITLE
sage dependency updates

### PR DESCRIPTION
**Pull request by collares**
I will slowly work on this.

### CODE REVIEW

**Comment by collares** [comment]
@ofborg build sage

File: pkgs/applications/science/math/pari/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/pari/default.nix:20]:
```suggestion
    # Old versions are at 
    urls = [
      "https://pari.math.u-bordeaux.fr/pub/pari/unix/${pname}-${version}.tar.gz";
      "https://pari.math.u-bordeaux.fr/pub/pari/OLD/${lib.versions.majorMinor version}/${pname}-${version}.tar.gz"
    ];
```

File: pkgs/development/libraries/arb/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/arb/default.nix:11]:
```suggestion
  buildInputs = [ mpir gmp mpfr flint ];
```

File: pkgs/development/libraries/arb/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/arb/default.nix:2]:
```suggestion
{lib, stdenv, fetchFromGitHub, mpir, gmp, mpfr, flint}:

stdenv.mkDerivation rec {
```

Comment (AndersonTorres) [inline pkgs/development/libraries/arb/default.nix:2]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, mpir
, gmp
, mpfr
, flint
}:

stdenv.mkDerivation rec {
```

Just a suggestion.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
@collares ping :)

File: pkgs/development/libraries/arb/default.nix
Comment (collares) [reply pkgs/development/libraries/arb/default.nix]: Done, thanks!

File: pkgs/applications/science/math/pari/default.nix
Comment (collares) [reply pkgs/applications/science/math/pari/default.nix]: Great idea, thanks! I've implemented it.

**Review by omasanori — COMMENTED** [review]
Thank you so much for your excellent work as usual!
Just a few minor comments for now; I will review several times for a better result.

**Comment by collares** [comment]
Many thanks for all the reviews everyone :) I value your comments but I also value your time, so (while I will quickly apply any suggested changes) I should clarify that nothing in this draft PR passes tests yet, and it might take a long time to debug the failures and get this to a mergeable state. Thanks for your patience!

File: pkgs/applications/science/math/palp/default.nix
Comment (collares) [reply pkgs/applications/science/math/palp/default.nix]: Done, thanks!

**Comment by omasanori** [comment]
FYI: https://trac.sagemath.org/ticket/30537

**Comment by symphorien** [comment]
xcas still works ☑

**Comment by collares** [comment]
@symphorien Thanks for checking! In a sense `giac` is the most annoying part, because the nofltk patch was not updated so we cannot run tests on fltk-less giac anymore (but we can on `giac-with-xcas`, which should be enough). I will look into updating the patch.

**Comment by collares** [comment]
@ofborg build giac-with-xcas

**Comment by collares** [comment]
Status update:

1. The cypari2 caused a new test failure on aarch64, and fixing it properly requires Cython changes (#115200). I've included that PR here for testing purposes, but I'll remove it once it gets merged to master (which will take a few weeks at best, since if approved it'll land in staging).
2. Given the above, this PR will take a couple of weeks to land, so I've decided to drop the pari update for now. Upstream still needs to update some algorithms related to Smith Normal Form changes, and those changes are causing test failures at the moment. I'll reinclude the pari update if it lands upstream before the Cython changes stabilize.
3. I've contacted giac developers about new test failures with fltk disabled, and I've disabled the failiing tests (`chk_fhan2` and `chk_fhan9`) for now.

**Comment by collares** [comment]
@omasanori Even though this does not update cypari2 or pari, I am marking this as ready for review. I will submit a PR to update cypari2 on staging as soon as this PR is merged, but will probably wait a little bit more for the pari upgrade. 

Note that I've included an upstream patch to fix numpy usage which currently breaks sage tests. If you don't have enough time to review the whole thing I can split the numpy fix off into a separate PR. I won't wait for ofborg's test run, but if it's green this should be ready to merge.

**Comment by omasanori** [comment]
Thank you so much, @collares! I will submit my review results by next weekend.

**Review by omasanori — COMMENTED** [review]
Could anyone test the following tests on top of this PR?
```shell
nix-build -E 'with (import ./. {}); sage.tests.override { files = [
  "src/sage/finance/time_series.pyx"
  "src/sage/geometry/polyhedron/library.py"
  "src/sage/geometry/polyhedron/base.py"
  "src/sage/matrix/matrix_double_dense.pyx"
  "src/sage/numerical/sdp.pyx"
  "src/sage/tests/books/computational-mathematics-with-sagemath/float_doctest.py"
]; }'
```
On my test setup some tests fail though they pass on Hydra. I will reinstall the setup next week but I want to get confidence that they are not reproducible failures before doing that.

Some minor nitpicking below.

**Comment by collares** [comment]
@omasanori That's intriguing! Those tests pass here. Can you paste the failing test output here or in a Gist?

**Comment by omasanori** [comment]
Yes, of course.

 - system: `"x86_64-linux"`
 - host os: `Linux 3.10.0-1160.11.1.el7.x86_64, CentOS Linux, 7 (Core)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(ogino): `"nixpkgs-21.05pre273941.a2b0ea6865b"`
 - nixpkgs: `/home/ogino/.nix-defexpr/channels/nixpkgs`

<details>
<summary>src/sage/finance/time_series.pyx</summary>
<pre><code>**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 608, in sage.finance.time_series.TimeSeries.autoregressive_fit
Failed example:
    v.autoregressive_forecast(F)
Expected:
    86.0177285042...
Got:
    243.17491459219573
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 610, in sage.finance.time_series.TimeSeries.autoregressive_fit
Failed example:
    F
Expected:
    [1.0148, -0.0029, -0.0105, 0.0067, -0.0232 ... -0.0106, -0.0068, 0.0085, -0.0131, 0.0092]
Got:
    [2.7845, -1.0730, 154.7546, -160.5283, 2.5521 ... -0.0583, 0.0665, -0.0665, 0.9782, -0.8029]
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 2650, in sage.finance.time_series.autoregressive_fit
Failed example:
    F = finance.autoregressive_fit(ac); F
Expected:
    [0.9982, -0.0002, -0.0002, 0.0003, 0.0001 ... 0.0002, -0.0002, -0.0000, -0.0002, -0.0014]
Got:
    [3810.5722, 2916.4368, -3042.5908, -189282.4881, 187717.7192 ... 0.0000, -0.0000, 0.0051, -0.4606, 0.4554]
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 2655, in sage.finance.time_series.autoregressive_fit
Failed example:
    sum(F)
Expected:
    0.99593284089454...
Got:
    8504.685577940327
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 2667, in sage.finance.time_series.autoregressive_fit
Failed example:
    y2[:-1].autoregressive_forecast(F)
Expected:
    6.7836741372407...
Got:
    128906.19937810324
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 2673, in sage.finance.time_series.autoregressive_fit
Failed example:
    y2[:-1].autoregressive_forecast(y2[:-1].autoregressive_fit(len(y2)))
Expected:
    6.770168705668...
Got:
    9.380452670879273e+29
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 2688, in sage.finance.time_series.autoregressive_fit
Failed example:
    s2/s1
Expected:
    1.15464636102...
Got:
    214.6708196021793
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/finance/time_series.pyx", line 2706, in sage.finance.time_series.autoregressive_fit
Failed example:
    s.mean()
Expected:
    1.354073591877...
Got:
    0.2404119161741636
**********************************************************************
2 items had failures:
   2 of  16 in sage.finance.time_series.TimeSeries.autoregressive_fit
   6 of  21 in sage.finance.time_series.autoregressive_fit
    [343 tests, 8 failures, 11.70 s]</code></pre>
</details>
<details>
<summary>src/sage/geometry/polyhedron/library.py</summary>
<pre><code>**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/geometry/polyhedron/library.py", line 2475, in sage.geometry.polyhedron.library.Polytopes.hypersimplex
Failed example:
    TestSuite(h_7_3).run(skip="_test_pyramid")
Expected nothing
Got:
    Failure in _test_linear_transformation:
    Traceback (most recent call last):
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/misc/sage_unittest.py", line 297, in run
        test_method(tester=tester)
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/base.py", line 5332, in _test_linear_transformation
        tester.assertEqual(self, self.linear_transformation(identity_matrix(self.ambient_dim())))
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/base.py", line 5309, in linear_transformation
        return new_parent.element_class(new_parent, [new_vertices, new_rays, new_lines],
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/backend_cdd.py", line 520, in __init__
        Polyhedron_cdd.__init__(self, parent, Vrep, Hrep, **kwds)
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/base.py", line 208, in __init__
        self._init_from_Vrepresentation_and_Hrepresentation(Vrep, Hrep)
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/backend_cdd.py", line 646, in _init_from_Vrepresentation_and_Hrepresentation
        try_init(prim)
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/backend_cdd.py", line 621, in try_init
        s = self._run_cdd(s, '--repall', verbose=verbose)
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/geometry/polyhedron/backend_cdd.py", line 170, in _run_cdd
        raise ValueError(ans.strip())
    ValueError: *Error: Numerical inconsistency is found.  Use the GMP exact arithmetic.
    ------------------------------------------------------------
    The following tests failed: _test_linear_transformation
**********************************************************************
1 item had failures:
   1 of   9 in sage.geometry.polyhedron.library.Polytopes.hypersimplex
    [307 tests, 1 failure, 82.48 s]</code></pre>
</details>
<details>
<summary>src/sage/geometry/polyhedron/base.py</summary>
<pre><code>**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/geometry/polyhedron/base.py", line 8007, in sage.geometry.polyhedron.base.Polyhedron_base.volume
Failed example:
    w = Dinexact.faces(2)[2].as_polyhedron().volume(measure='induced', engine='internal'); RDF(w) # abs tol 1e-9
Expected:
    1.5340627082974878
Got:
    8.032456584010367
Tolerance exceeded:
    1.5340627082974878 vs 8.032456584010367, tolerance 7e0 > 1e-9
**********************************************************************
1 item had failures:
   1 of  46 in sage.geometry.polyhedron.base.Polyhedron_base.volume
    [1575 tests, 1 failure, 83.19 s]</code></pre>
</details>
<details>
<summary>src/sage/matrix/matrix_double_dense.pyx</summary>
<pre><code>**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/matrix/matrix_double_dense.pyx", line 625, in sage.matrix.matrix_double_dense.Matrix_double_dense.condition
Failed example:
    A.condition()  # tol 2e-4
Expected:
    16332197709146.014
Got:
    276341912583183.9
Tolerance exceeded:
    16332197709146.014 vs 276341912583183.9, tolerance 2e1 > 2e-4
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/matrix/matrix_double_dense.pyx", line 1347, in sage.matrix.matrix_double_dense.Matrix_double_dense.?
Failed example:
    A.eigenvalues(algorithm='symmetric', tol=1.0e-5)  # tol 2e-15
Expected:
    [(-8.0, 22), (2.0, 77), (22.0, 1)]
Got:
    [(-13.81753974166025, 1),
     (-13.648124559508018, 1),
     (-12.020844694797459, 1),
     (-11.501673730831273, 1),
     (-10.996735991707883, 1),
     (-10.034951645118767, 1),
     (-9.962283093579666, 1),
     (-9.128876066618123, 1),
     (-8.996148594502737, 1),
     (-8.722692543787009, 1),
     (-8.604556296665711, 1),
     (-8.40884373343103, 1),
     (-8.374989428968158, 1),
     (-8.04552871933626, 1),
     (-8.0, 13),
     (-7.526372396375868, 1),
     (-6.831403963944408, 1),
     (-6.629587466119391, 1),
     (-6.4982355637201605, 1),
     (-6.265676025803857, 1),
     (-6.234896753104995, 1),
     (-5.716331549376318, 1),
     (-5.390377899637494, 1),
     (-5.264189370979771, 1),
     (-4.717055974115992, 1),
     (-4.342049462991852, 1),
     (-4.035470777620305, 1),
     (-3.759935252697173, 1),
     (-3.3078409444345485, 1),
     (-2.9815285949726116, 1),
     (-2.4404128190223333, 1),
     (-1.8924099853815357, 1),
     (-1.844087730193682, 1),
     (-1.4442736849343316, 1),
     (-0.9294180296023848, 1),
     (-0.8110452755504824, 1),
     (-0.5222794367105574, 1),
     (0.20873807157486873, 1),
     (0.3951954399812892, 1),
     (0.8759590236699589, 1),
     (1.4381479600752756, 1),
     (1.532163686397235, 1),
     (1.7434392786321635, 1),
     (1.8865575972656636, 1),
     (2.0000000193635405, 14),
     (2.265318402262459, 1),
     (2.5658867016601037, 1),
     (2.817713598035522, 1),
     (3.081870153164812, 1),
     (3.446850975471049, 1),
     (3.6512306004984585, 1),
     (3.878710063182693, 1),
     (4.014287691943725, 1),
     (4.360367867192431, 1),
     (4.7229340629669, 1),
     (4.982801964690831, 1),
     (5.37818178439969, 1),
     (6.080935570154745, 1),
     (6.348181367606261, 1),
     (6.525599712412862, 1),
     (6.915551936720855, 1),
     (7.3044283658318365, 1),
     (7.452144258588678, 1),
     (7.5567933585381555, 1),
     (7.8716696840779745, 1),
     (7.937771590598039, 1),
     (8.114444731912405, 1),
     (8.195204412460074, 1),
     (8.780171598716006, 1),
     (10.615868120675568, 1),
     (12.063743301519922, 1),
     (12.935140199920008, 1),
     (12.961615404888553, 1),
     (17.06551630288574, 1),
     (21.999999999999986, 1)]
**********************************************************************
2 items had failures:
   1 of  49 in sage.matrix.matrix_double_dense.Matrix_double_dense.?
   1 of  42 in sage.matrix.matrix_double_dense.Matrix_double_dense.condition
    [629 tests, 2 failures, 2.69 s]</code></pre>
</details>
<details>
<summary>src/sage/numerical/sdp.pyx</summary>
<pre><code>**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/numerical/sdp.pyx", line 125, in sage.numerical.sdp
Failed example:
    x=p.solve(); x+1
Exception raised:
    Traceback (most recent call last):
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/doctest/forker.py", line 720, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/nix/store/9kjg3zzx8rfd3vs716scs2f4k8fx4qvg-python3-3.8.8-env/lib/python3.8/site-packages/sage/doctest/forker.py", line 1153, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.numerical.sdp[28]>", line 1, in <module>
        x=p.solve(); x+Integer(1)
      File "sage/numerical/sdp.pyx", line 964, in sage.numerical.sdp.SemidefiniteProgram.solve (build/cythonized/sage/numerical/sdp.c:8297)
        self._backend.solve()
      File "sage/numerical/backends/cvxopt_sdp_backend.pyx", line 160, in sage.numerical.backends.cvxopt_sdp_backend.CVXOPTSDPBackend.solve (build/cythonized/sage/numerical/backends/cvxopt_sdp_backend.c:3423)
        raise SDPSolverException("CVXOPT: Terminated early due to numerical difficulties or because the maximum number of iterations was reached.")
    sage.numerical.sdp.SDPSolverException: CVXOPT: Terminated early due to numerical difficulties or because the maximum number of iterations was reached.
**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/numerical/sdp.pyx", line 130, in sage.numerical.sdp
Failed example:
    p.slack(0).trace()        # tol 1e-14
Expected:
    1.0
Got:
    -0.0550587074717638
Tolerance exceeded:
    1.0 vs -0.0550587074717638, tolerance 2e0 > 1e-14
**********************************************************************
1 item had failures:
   2 of  46 in sage.numerical.sdp
    [288 tests, 2 failures, 0.88 s]</code></pre>
</details>
<details>
<summary>src/sage/tests/books/computational-mathematics-with-sagemath/float_doctest.py</summary>
<pre><code>**********************************************************************
File "/nix/store/pqarg7pym821rdg3bnyvmkmn59xh8vjp-sage-src-9.2/src/sage/tests/books/computational-mathematics-with-sagemath/float_doctest.py", line 402, in sage.tests.books.computational-mathematics-with-sagemath.float_doctest
Failed example:
    NearlySingularMatrix(RDF,n).det() == 0.0
Expected:
    True
Got:
    False
**********************************************************************
1 item had failures:
   1 of 142 in sage.tests.books.computational-mathematics-with-sagemath.float_doctest
    [141 tests, 1 failure, 40.11 s]</code></pre>
</details>

Full log: https://gist.github.com/omasanori/fb150eca7e3261118eebff56fe8d393c

**Comment by symphorien** [comment]
Heads up: with a single user installation, the store is actually readwrite, and python likes to corrupt the store by compiling pyc files inside it. This can cause strange side effects.

File: pkgs/applications/science/math/giac/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/giac/default.nix:79]:
```suggestion
    touch check/chk_fhan2
    touch check/chk_fhan9
```

File: pkgs/applications/science/math/giac/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/giac/default.nix:44]: Please use substituteInPlace instead.

File: pkgs/applications/science/math/giac/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/giac/default.nix:9]:
```suggestion
```

those packages should never be null.

File: pkgs/applications/science/math/giac/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/giac/default.nix:5]:
```suggestion
, enableGUI ? false, enableMicroPy ? false, python3
, libGL, libGLU, xorg, fltk
```

File: pkgs/applications/science/math/giac/default.nix
Comment (symphorien) [reply pkgs/applications/science/math/giac/default.nix:79]: nope this is used to *empty* the file

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 114881` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
14 packages built:
</summary>
<ul>
<li>arb</li>
<li>giac</li>
<li>giac-with-xcas</li>
<li>jmol</li>
<li>palp</li>
<li>python38Packages.cypari2</li>
<li>python38Packages.cysignals</li>
<li>python38Packages.fpylll</li>
<li>python38Packages.pplpy</li>
<li>python39Packages.cypari2</li>
<li>python39Packages.cysignals</li>
<li>python39Packages.fpylll</li>
<li>python39Packages.pplpy</li>
<li>sympow</li>
</ul>
</details>

File: pkgs/applications/science/math/giac/default.nix
Comment (symphorien) [reply pkgs/applications/science/math/giac/default.nix:44]: substituteInPlace does no support regexps

File: pkgs/applications/science/math/giac/default.nix
Comment (symphorien) [reply pkgs/applications/science/math/giac/default.nix:9]: Your dislike of null is not universal, and I think that this is a bit too opinionated of a nitpick on a quite difficult PR. Can you let the author concentrate on the actual issues at stake please?

Comment (collares) [reply pkgs/applications/science/math/giac/default.nix:9]: I was just following the existing package convention of not depending on packages that are only conditionally needed. I can definitely see the argument that Python 3 is ubiquitous while "xorg" might not be available on certain platforms, though. I will let @symphorien decide though, since he's this package's maintainer.

File: pkgs/applications/science/math/giac/default.nix
Comment (collares) [reply pkgs/applications/science/math/giac/default.nix:5]: Thanks for the suggestion, but this one might be bad for Darwin users that don't want to pull in a xorg dependency for a command-line utility.

**Comment by collares** [comment]
> Heads up: with a single user installation, the store is actually readwrite, and python likes to corrupt the store by compiling pyc files inside it. This can cause strange side effects.

Would `nix-store --repair --verify --check-contents` fix that?

File: pkgs/applications/science/math/giac/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/science/math/giac/default.nix:9]: > Your dislike of null is not universal

``? null`` for optional features it is not required and even discouraged because hides errors when renaming and removing packages and breaks overrides in certain cases. There are certain use cases where it is required but I am pretty sure in this case it will not cause any regressions and make sage easier to maintain.

> I was just following the existing package convention of not depending on packages that are only conditionally needed.

nixpkgs goes the other way most of the time and just includes most optional features by default.

> "xorg" might not be available on certain platforms

even if it is, it will never be null and marked broken or platforms will not allow it.

If an input package is null this is a bug and should be investigated.

Comment (collares) [reply pkgs/applications/science/math/giac/default.nix:9]: Fair enough. I don't have a horse in this race, so I will let the giac package maintainer decide if removing `? null` is a desirable change. (By the way, is there a treewide effort to remove `? null` in such cases? I ask because knowing of such an effort would make the decision trivial.)

@symphorien What do you think? I don't mind at all doing an extra rebase. @SuperSandro2000 already changed it in #115361, which I think is an overstepping, but I urge you to consider this merely on technical grounds.

My opinion (which I am **not** claiming matters) is that users should not need to download `fltk` or `xorg.libX11` if they want to use `giac` and not `giac-with-xcas`, and if removing `? null` does keep this property (and does not break Darwin) I don't really mind.

**Review by omasanori — COMMENTED** [review]
> Heads up: with a single user installation, the store is actually readwrite, and python likes to corrupt the store by compiling pyc files inside it. This can cause strange side effects.

Thank you @symphorien, I did not know this.

Anyway, the failures I am facing occur with or without changes on this branch, so it does not introduce regression on my machine, at least. I have no objections and will approve after rebasing to resolve conflicts.

**Comment by collares** [comment]
@omasanori Rebased. I was going to wait for the cysignals update but I don't think it matters much since I had already imported the relevant patch. So I think this is ready if it passes tests.

File: pkgs/applications/science/math/giac/default.nix
Comment (symphorien) [reply pkgs/applications/science/math/giac/default.nix]: My point is not on the technical merits of adding or removing `null`, just that this PR has enough meat that such debatable nitpicking might be detrimental. But I guess my comment was counterproductive, so let's cut it short: do as you wish.

Comment (collares) [reply pkgs/applications/science/math/giac/default.nix]: @symphorien I appreciate you stepping in to make sure the PR wasn't derailed by the `null` point, thank you! I offered to do an extra rebase for the same reason, in fact.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 114881` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
14 packages built:
</summary>
<ul>
<li>arb</li>
<li>giac</li>
<li>giac-with-xcas</li>
<li>palp</li>
<li>python38Packages.cypari2</li>
<li>python38Packages.cysignals</li>
<li>python38Packages.fpylll</li>
<li>python38Packages.pplpy</li>
<li>python39Packages.cypari2</li>
<li>python39Packages.cysignals</li>
<li>python39Packages.fpylll</li>
<li>python39Packages.pplpy</li>
<li>rankwidth</li>
<li>sympow</li>
</ul>
</details>

**Review by omasanori — APPROVED** [review]
Thank you all so much for your invaluable help!

**Comment by collares** [comment]
@omasanori You're welcome! Thanks for the review. I'll make another PR updating `cypari2`, `cysignals`, `eclib` and `pari` as soon as upstream finishes up the pari upgrade.

**Comment by symphorien** [comment]
> So I think this is ready if it passes tests.

it passes tests so let me merge.

**Comment by timokau** [comment]
Thank you for your efforts to keep the sage package running and up to date :)

**Pull request merged**
